1. promise 可以用于异步操作（异步操作意味着同时进行），并且能判断出是出现错误了还是返回了需要的结果；

> 单线程怎么会同时进行？？？即便有任务队列，回调函数也要有地方执行啊？？？，待填坑

2. promise 的设计理念及实现方式是很好的，在只有一个 promise 的时候最好，promise 一多，再加上如果需要传参，代码也会不好理解；

3. async-await登场了。

4. async 函数本身返回一个 promise，这是极好的，可以发挥一个 promise 的优点：异步调用，并且可以判断是否返回期许的数据；

5. await 必须在 async 函数中使用，后面可以是任何表达式，但他会等一个结果，或者说返回一个结果（直接量，或者如果他后面是 promise 的时候处理成直接量）；

6. 如果 await 后面是一个 promise（只有是 promise 的情况下才能发挥其优势，一般是另一个 async 函数的返回值，或者像 fetch 这种封装好的返回 promise 的API），await 要做的工作比较多了：如果 promise resolve 了，那么他会拿到 rosolve 的值，如果 reject 抛异常，所以最好放在 try-catch 中。    

7. 还有一个，async 函数和 promise 的执行器一样，都会理解执行，但，async 函数里面，如果遇到 await，便会阻塞，必须等到 await 等到一个结果才能继续执行函数内部后面的代码。

```js
async function foo() {
  let b = 0;
  for (let a = 1; a < 10000; a++ ) {
    b = b + a;
  }

  return b;
}

async function test() {
  console.log(11);

  // 想模拟请求延时，但 foo 函数写的不好
  const a = await foo();
  console.log(a);
  console.log(333)
}

test();
console.log(444);
```

NOTE 1：如果是定时器，不会阻塞，定时器 确实是个神奇的东西。

```js
async function test() {
  console.log(111);
  const a = await setTimeout(() => console.log(222), 1000);  
  console.log(a);
  console.log(333)
}

test();
console.log(444);
```

8. 其他：promise 的执行器是立即执行的，resolve 和 reject 的存在，是其异步操作的关键：   

```js
function testP() {
  let p = new Promise((resolve, reject) => {
    console.log(111);
    setTimeout(() => {
      resolve(222);
    }, 1000);
  })

  return p.then((v) => console.log(v));
}

testP();

console.log(333);
```